google/cloud/spanner/internal/connection_impl.cc

```diff
@@ -512,11 +512,10 @@ StatusOr<ResultType> ConnectionImpl::CommonDmlImpl(
   if (!prepare_status.ok()) {
     return prepare_status;
   }
-  // Capture a copy of of these member variables to ensure the `shared_ptr<>`
-  // remains valid through the lifetime of the lambda. Note that the local
-  // variables are a reference to avoid increasing refcounts twice, but the
-  // capture is by value.
-  auto const& stub = stub_;
+  // Capture a copy of of these to ensure the `shared_ptr<>` remains valid
+  // through the lifetime of the lambda. Note that the local variables are a
+  // reference to avoid increasing refcounts twice, but the capture is by value.
+  auto const& stub = session->stub();
```

Comment:
> FYI, I suspect the const& here does nothing now; I was thinking something like the compiler might see that it's doing lifetime extension and can effectively std::move it into the lambda capture, but I guess if it's smart enough to figure that out it could also figure it out if it was just a value. So, should I delete it?

Yes, I don't think this is doing anything different to `auto const stub = session->stub()`.  You still need to give the stub a name though, so that `retry_resume_fn` can capture it.

> Is there any way to accomplish the goal of not doing an extra ref/unref other than having Session return a shared_ptr<>&? I didn't want to do that because it seems dangerous.

I don't think so, as the refcount has already gone up before `session->stub()` returns by value.  As to whether it is "dangerous" or not (I assume you meant `Session::stub()`), I'll say that depends on how you use it.  It seems OK to me, as long as you're done a `PrepareSession(session)` "_before_" so that `session->stub_` is const.

> Does just doing session->stub()->Call() now bump the reference count as well or can that be optimized away?

I would say it does incr/decr the refcount.  There is nothing, for example, that might stop a concurrent "`session->clear_stub()`" call from happening between the return of `stub()` and the call of `Call()`.